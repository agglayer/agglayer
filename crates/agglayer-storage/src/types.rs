use agglayer_types::{Certificate, CertificateHeader, CertificateId, Digest, NetworkId, Proof};
use serde::{Deserialize, Serialize};

use crate::columns::Codec;

macro_rules! default_codec_impl {
    ($($ident: ident),+) => {
        $(impl crate::columns::Codec for $ident {})+
    };
}

#[derive(Debug, Serialize, Deserialize)]
pub enum MetadataKey {
    LatestSettledEpoch,
    EpochSynchronization,
}

#[derive(Debug, Serialize, Deserialize)]
pub enum MetadataValue {
    LatestSettledEpoch(u64),
    EpochSynchronization(u64),
}

#[derive(Debug, Serialize, Deserialize)]
pub enum PerEpochMetadataKey {
    SettlementTxHash,
    Packed,
}

#[derive(Debug, Serialize, Deserialize)]
pub enum PerEpochMetadataValue {
    SettlementTxHash(Digest),
    Packed(bool),
}

#[derive(Debug, Serialize, Deserialize, PartialEq, Eq, PartialOrd, Ord)]
pub struct SmtKey {
    pub(crate) network_id: u32,
    pub(crate) key_type: SmtKeyType,
}

#[derive(Clone, Debug, Serialize, Deserialize, PartialEq, Eq, PartialOrd, Ord)]
pub enum SmtKeyType {
    Root,
    Node(Digest),
}

#[derive(Clone, PartialEq, Eq, Debug, Serialize, Deserialize)]
pub enum SmtValue {
    Node(Digest, Digest),
    Leaf(Digest),
}

impl Codec for SmtKey {}
impl Codec for SmtValue {}

default_codec_impl!(
    u64,
    u32,
    Certificate,
    CertificateId,
    CertificateHeader,
    MetadataKey,
    MetadataValue,
    NetworkId,
    PerEpochMetadataKey,
    PerEpochMetadataValue,
    Proof
);
