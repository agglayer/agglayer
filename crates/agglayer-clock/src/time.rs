use std::{
    num::NonZeroU64,
    sync::{
        atomic::{AtomicU64, Ordering},
        Arc,
    },
    time::Duration,
};

use chrono::{DateTime, Utc};
use tokio::{
    sync::broadcast,
    time::{interval_at, Instant},
};
use tokio_util::sync::CancellationToken;
use tracing::{debug, error};

use crate::{Clock, ClockRef, Error, Event, BROADCAST_CHANNEL_SIZE};

/// Time based [`Clock`] implementation.
///
/// Simulate blockchain block production by increasing Block height by 1 every
/// second. Epoch duration can be configured when creating the Clock.
pub struct TimeClock {
    genesis: DateTime<Utc>,
    current_block: Arc<AtomicU64>,
    epoch_duration: NonZeroU64,
    current_epoch: Arc<AtomicU64>,
}

#[async_trait::async_trait]
impl Clock for TimeClock {
    async fn spawn(mut self, cancellation_token: CancellationToken) -> Result<ClockRef, Error> {
        let (sender, _receiver) = broadcast::channel(BROADCAST_CHANNEL_SIZE);

        let clock_ref = ClockRef {
            sender: sender.clone(),
            current_epoch: self.current_epoch.clone(),
            block_height: self.current_block.clone(),
        };

        // Spawn the Clock task directly
        tokio::spawn(async move {
            self.run(sender, cancellation_token).await;
        });

        Ok(clock_ref)
    }
}

impl TimeClock {
    /// Create a new [`TimeClock`] instance based on the current datetime and an
    /// Epoch.
    pub fn new_now(epoch_duration: NonZeroU64) -> Self {
        Self::new(Utc::now(), epoch_duration)
    }

    /// Create a new [`TimeClock`] instance based on a genesis datetime and an
    /// Epoch duration.
    pub fn new(genesis: DateTime<Utc>, epoch_duration: NonZeroU64) -> Self {
        Self {
            genesis,
            current_block: Arc::new(AtomicU64::new(0)),
            epoch_duration,
            current_epoch: Arc::new(AtomicU64::new(0)),
        }
    }

    /// Run the Clock task.
    async fn run(
        &mut self,
        sender: broadcast::Sender<Event>,
        cancellation_token: CancellationToken,
    ) {
        let mut interval = interval_at(Instant::now(), Duration::from_secs(1));

        // Compute the current Block height and Epoch number
        let current_block = self.update_block_height();
        let current_epoch = self.calculate_epoch_number(current_block);

        // Use compare_exchange to ensure the initial current_epoch is 0
        if let Err(stored_value) = self.current_epoch.compare_exchange(
            0,
            current_epoch,
            Ordering::AcqRel,
            Ordering::Relaxed,
        ) {
            let error_message = format!(
                "The current_epoch has already been modified. Shutting down the Clock task. \
                 Stored value: {}",
                stored_value
            );
            #[cfg(not(test))]
            {
                error!("{}", error_message);
                std::process::exit(1);
            }
            #[cfg(test)]
            panic!("{}", error_message);
        }

        loop {
            tokio::select! {
                _ = cancellation_token.cancelled() => {
                    debug!("Clock task cancelled");
                    break;
                }
                _ = interval.tick() => {

                    // Increase the Block height by 1.
                    // The `fetch_add` method returns the previous value, so we need to add 1 to it
                    // to get the current Block height.
                    if let Some(current_block) = self
                        .current_block
                            .fetch_add(1, Ordering::Release)
                            .checked_add(1)
                    {
                        // If the current Block height is a multiple of the Epoch duration,
                        // the current Epoch has ended. In this case, we need to update the
                        // new Epoch number and send an `EpochEnded` event to the subscribers.
                        if current_block % self.epoch_duration == 0 {
                            match self.update_epoch_number() {
                                Ok(epoch_ended) => {
                                    _ = sender.send(Event::EpochEnded(epoch_ended));
                                }
                                Err((current_epoch, expected)) => {
                                    error!(
                                        "Unexpected error computing the current Epoch: current_epoch={}, \
                                        expected_epoch={}, current_block={}",
                                        current_epoch, expected, current_block
                                    );
                                    cancellation_token.cancel();
                                }
                            }
                        }
                    } else {
                       error!("Block height overflowed the u64 limit. \
                           This is an unexpected situation and could lead to unexpected behavior. \
                           Please report this issue to the developers. https://github.com/agglayer/agglayer/issues/new \
                           The node will now kill itself to prevent further damage.");

                        cancellation_token.cancel();
                    }
                }
            }
        }
    }

    /// Updates the Block height of this [`TimeClock`].
    ///
    /// This method is used to update the Block height based on the
    /// genesis datetime and the current datetime.
    ///
    /// The Block height is the number of seconds since the genesis datetime.
    fn update_block_height(&mut self) -> u64 {
        let block_height = self.calculate_block_height();

        self.current_block.store(block_height, Ordering::Release);

        block_height
    }

    /// Updates the current Epoch of this [`TimeClock`].
    ///
    /// This method is used to update the current Epoch number based on the
    /// Block height and the Epoch duration.
    ///
    /// To define the current Epoch number, the Epoch duration divides the Block
    /// height.
    fn update_epoch_number(&mut self) -> Result<u64, (u64, u64)> {
        let current_block = self.current_block.load(Ordering::Acquire);

        let current_epoch = self.calculate_epoch_number(current_block);
        let expected_epoch = current_epoch.saturating_sub(1);

        match self.current_epoch.compare_exchange(
            expected_epoch,
            current_epoch,
            Ordering::Acquire,
            Ordering::Relaxed,
        ) {
            Ok(previous) => Ok(previous),
            Err(stored) => Err((stored, expected_epoch)),
        }
    }

    /// Calculate an Epoch number based on a Block number.
    fn calculate_epoch_number(&self, from_block: u64) -> u64 {
        from_block / self.epoch_duration
    }

    /// Calculate the Block height.
    fn calculate_block_height(&self) -> u64 {
        std::cmp::max(
            Utc::now()
                .naive_utc()
                .signed_duration_since(self.genesis.naive_utc())
                .num_seconds(),
            0,
        ) as u64
    }
}

#[cfg(test)]
mod tests {
    use std::{num::NonZeroU64, sync::atomic::Ordering};

    use chrono::{Duration, Utc};
    use tokio::sync::broadcast;
    use tokio_util::sync::CancellationToken;

    use crate::{Clock, ClockRef, Event, TimeClock, BROADCAST_CHANNEL_SIZE};

    #[tokio::test]
    async fn test_time_clock() {
        let genesis = Utc::now()
            .checked_sub_signed(Duration::seconds(30))
            .unwrap();

        let clock = TimeClock::new(genesis, NonZeroU64::new(5).unwrap());

        let token = CancellationToken::new();
        let clock_ref = clock.spawn(token.clone()).await.unwrap();

        let mut recv = clock_ref.subscribe().unwrap();

        assert_eq!(recv.recv().await, Ok(Event::EpochEnded(6)));
        assert_eq!(clock_ref.current_epoch(), 7);
        assert!(clock_ref.current_block_height() >= 30);
    }

    #[tokio::test]
    async fn test_time_clock_catchup() {
        let genesis = Utc::now()
            .checked_sub_signed(Duration::seconds(30))
            .unwrap();

        let clock = TimeClock::new(genesis, NonZeroU64::new(2).unwrap());

        let token = CancellationToken::new();
        let clock_ref = clock.spawn(token.clone()).await.unwrap();

        let mut recv = clock_ref.subscribe().unwrap();
        assert_eq!(recv.recv().await, Ok(Event::EpochEnded(15)));
        assert!(recv.try_recv().is_err());
        assert_eq!(clock_ref.current_epoch(), 16);
        assert!(clock_ref.current_block_height() >= 30);
        tokio::time::sleep(std::time::Duration::from_secs(5)).await;

        assert_eq!(recv.recv().await, Ok(Event::EpochEnded(16)));
        assert_eq!(recv.recv().await, Ok(Event::EpochEnded(17)));

        assert_eq!(clock_ref.current_epoch(), 18);
        assert!(clock_ref.current_block_height() >= 35);
    }

    #[tokio::test]
    async fn test_time_clock_overflow() {
        let genesis = Utc::now()
            .checked_sub_signed(Duration::seconds(30))
            .unwrap();

        let mut clock = TimeClock::new(genesis, NonZeroU64::new(2).unwrap());
        let blocks = clock.current_block.clone();
        let (sender, _receiver) = broadcast::channel(BROADCAST_CHANNEL_SIZE);

        let clock_ref = ClockRef {
            sender: sender.clone(),
            current_epoch: clock.current_epoch.clone(),
            block_height: clock.current_block.clone(),
        };

        let token = CancellationToken::new();
        let mut fut = Box::pin(clock.run(sender, token.clone()));
        let mut recv = clock_ref.subscribe().unwrap();

        _ = futures::poll!(&mut fut);
        tokio::time::sleep(std::time::Duration::from_secs(2)).await;

        _ = futures::poll!(&mut fut);
        assert_eq!(recv.try_recv(), Ok(Event::EpochEnded(15)));

        // Overflow the block height on next poll
        blocks.store(u64::MAX - 1, std::sync::atomic::Ordering::SeqCst);

        tokio::time::sleep(std::time::Duration::from_secs(2)).await;
        _ = futures::poll!(&mut fut);

        assert!(token.is_cancelled());
    }

    #[tokio::test]
    #[should_panic]
    async fn test_initial_epoch_update_error() {
        let genesis = Utc::now()
            .checked_sub_signed(Duration::seconds(30))
            .unwrap();

        let mut clock = TimeClock::new(genesis, NonZeroU64::new(5).unwrap());
        clock.current_epoch.store(1, Ordering::Relaxed);

        let (sender, _receiver) = broadcast::channel(BROADCAST_CHANNEL_SIZE);

        let token = CancellationToken::new();
        let _ = clock.run(sender, token).await;
    }
}
