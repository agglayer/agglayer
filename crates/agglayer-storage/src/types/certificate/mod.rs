//! Definitions of the certificate storage format with backwards compatibility.
//!
//! Currently, we have two versions of certificate storage format. The first byte determines the
//! storage format version.
//!
//! In version 0, where backwards compatibility is required, the first byte happens to be the
//! highest byte of network ID. This effectively limits the range of network IDs in v0 storage
//! to [0, 2^24-1]. The current network IDs fall into this range, so the highest byte (with value 0)
//! also acts as the version tag.
//!
//! In subsequent versions, we just have the version byte followed by a straightforward encoding of
//! the certificate, restoring the full range of network IDs.
//!
//! In the unlikely scenario where it turns out we need more than 256 storage format versions,
//! another byte can be allocated to specify a "sub-version" in one of the future versions.

use std::borrow::Cow;

use agglayer_types::{Certificate, Digest, Height, Metadata, NetworkId, Signature};
use bincode::Options;
use pessimistic_proof::{
    aggchain_proof::{AggchainProof, AggchainProofSP1},
    bridge_exit::BridgeExit,
    imported_bridge_exit::ImportedBridgeExit,
};
use serde::{Deserialize, Serialize};

use crate::columns::{default_bincode_options, CodecError};

/// A unit type serializing to a constant byte representing the storage version.
#[derive(Debug, Clone, Copy, Default, Serialize, Deserialize)]
#[serde(try_from = "u8", into = "u8")]
struct VersionTag<const VERSION: u8>;

impl<const VERSION: u8> TryFrom<u8> for VersionTag<VERSION> {
    type Error = CodecError;
    fn try_from(byte: u8) -> Result<Self, Self::Error> {
        (byte == VERSION)
            .then_some(Self)
            .ok_or(CodecError::BadCertificateVersion { version: byte })
    }
}

impl<const VERSION: u8> From<VersionTag<VERSION>> for u8 {
    fn from(VersionTag: VersionTag<VERSION>) -> Self {
        VERSION
    }
}

/// A three-byte network ID used in v0.
///
/// In v0, the first byte of network ID was reserved to specify the storage format version.
#[derive(Debug, Clone, Copy, Deserialize)]
#[cfg_attr(test, derive(Serialize))]
struct NetworkIdV0([u8; 3]);

impl From<NetworkIdV0> for NetworkId {
    fn from(NetworkIdV0([b2, b1, b0]): NetworkIdV0) -> NetworkId {
        u32::from_be_bytes([0, b2, b1, b0]).into()
    }
}

/// The pre-0.3 certificate format (`v0`).
#[derive(Debug, Clone, Deserialize)]
#[cfg_attr(test, derive(Serialize))]
struct CertificateV0 {
    version: VersionTag<0>,
    network_id: NetworkIdV0,
    height: Height,
    prev_local_exit_root: Digest,
    new_local_exit_root: Digest,
    bridge_exits: Vec<BridgeExit>,
    imported_bridge_exits: Vec<ImportedBridgeExit>,
    signature: Signature,
    metadata: Metadata,
}

impl From<CertificateV0> for Certificate {
    fn from(certificate: CertificateV0) -> Self {
        let CertificateV0 {
            version: VersionTag,
            network_id,
            height,
            prev_local_exit_root,
            new_local_exit_root,
            bridge_exits,
            imported_bridge_exits,
            signature,
            metadata,
        } = certificate;

        Certificate {
            network_id: network_id.into(),
            height,
            prev_local_exit_root,
            new_local_exit_root,
            bridge_exits,
            imported_bridge_exits,
            aggchain_proof: AggchainProof::ECDSA { signature },
            metadata,
        }
    }
}

/// The new certificate format as stored in the database (`v1`).
#[derive(Debug, Clone, Serialize, Deserialize)]
struct CertificateV1<'a> {
    version: VersionTag<1>,
    network_id: NetworkId,
    height: Height,
    prev_local_exit_root: Digest,
    new_local_exit_root: Digest,
    bridge_exits: Cow<'a, [BridgeExit]>,
    imported_bridge_exits: Cow<'a, [ImportedBridgeExit]>,
    aggchain_proof: AggchainProofV1<'a>,
    metadata: Metadata,
}

impl From<CertificateV1<'_>> for Certificate {
    fn from(certificate: CertificateV1) -> Self {
        let CertificateV1 {
            version: VersionTag,
            network_id,
            height,
            prev_local_exit_root,
            new_local_exit_root,
            bridge_exits,
            imported_bridge_exits,
            aggchain_proof,
            metadata,
        } = certificate;

        Certificate {
            network_id,
            height,
            prev_local_exit_root,
            new_local_exit_root,
            bridge_exits: bridge_exits.into_owned(),
            imported_bridge_exits: imported_bridge_exits.into_owned(),
            metadata,
            aggchain_proof: aggchain_proof.into(),
        }
    }
}

impl<'a> From<&'a Certificate> for CertificateV1<'a> {
    fn from(certificate: &'a Certificate) -> Self {
        let Certificate {
            network_id,
            height,
            prev_local_exit_root,
            new_local_exit_root,
            bridge_exits,
            imported_bridge_exits,
            metadata,
            aggchain_proof,
        } = certificate;

        CertificateV1 {
            version: VersionTag,
            network_id: *network_id,
            height: *height,
            prev_local_exit_root: *prev_local_exit_root,
            new_local_exit_root: *new_local_exit_root,
            bridge_exits: bridge_exits.into(),
            imported_bridge_exits: imported_bridge_exits.into(),
            aggchain_proof: aggchain_proof.into(),
            metadata: *metadata,
        }
    }
}

// Duplicated from `agglayer-types` since we need slightly different serde impls.
#[derive(Serialize, Deserialize, Clone, Debug)]
#[allow(clippy::upper_case_acronyms)]
pub enum AggchainProofV1<'a> {
    ECDSA {
        signature: Signature,
    },
    SP1 {
        aggchain_proof: Cow<'a, AggchainProofSP1>,
    },
}

impl<'a> From<&'a AggchainProof> for AggchainProofV1<'a> {
    fn from(proof: &'a AggchainProof) -> Self {
        match proof {
            AggchainProof::ECDSA { signature } => Self::ECDSA {
                signature: *signature,
            },
            AggchainProof::SP1 { aggchain_proof } => Self::SP1 {
                aggchain_proof: Cow::Borrowed(aggchain_proof),
            },
        }
    }
}

impl From<AggchainProofV1<'_>> for AggchainProof {
    fn from(proof: AggchainProofV1) -> Self {
        match proof {
            AggchainProofV1::ECDSA { signature } => Self::ECDSA { signature },
            AggchainProofV1::SP1 { aggchain_proof } => Self::SP1 {
                aggchain_proof: aggchain_proof.into_owned(),
            },
        }
    }
}

/// Type specifying the current certificate encoding format.
type CurrentCertificate<'a> = CertificateV1<'a>;

fn decode<T: for<'de> Deserialize<'de> + Into<Certificate>>(
    bytes: &[u8],
) -> Result<Certificate, CodecError> {
    Ok(default_bincode_options().deserialize::<T>(bytes)?.into())
}

impl crate::columns::Codec for Certificate {
    fn encode(&self) -> Result<Vec<u8>, CodecError> {
        // TODO get rid of the clones <https://github.com/agglayer/agglayer/issues/618>
        Ok(default_bincode_options().serialize(&CurrentCertificate::from(self))?)
    }

    fn decode(bytes: &[u8]) -> Result<Self, CodecError> {
        match bytes.first().copied() {
            None => Err(CodecError::CertificateEmpty),
            Some(0) => decode::<CertificateV0>(bytes),
            Some(1) => decode::<CertificateV1>(bytes),
            Some(version) => Err(CodecError::BadCertificateVersion { version }),
        }
    }
}

#[cfg(test)]
mod tests;
