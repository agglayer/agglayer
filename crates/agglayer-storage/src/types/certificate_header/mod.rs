use agglayer_types::{
    CertificateHeader, CertificateId, CertificateIndex, CertificateStatus, Digest, EpochNumber,
    Height, Metadata, NetworkId,
};
use bincode::Options as _;
use serde::Deserialize;

use crate::columns::{default_bincode_options, CodecError};

/// The pre-0.3 certificate format (`v0`).
#[derive(Debug, Clone, Deserialize)]
#[cfg_attr(test, derive(Serialize))]
struct CertificateHeaderV0 {
    network_id: NetworkId,
    height: Height,
    epoch_number: Option<EpochNumber>,
    certificate_index: Option<CertificateIndex>,
    certificate_id: CertificateId,
    prev_local_exit_root: Digest,
    new_local_exit_root: Digest,
    metadata: Metadata,
    status: CertificateStatus,
    settlement_tx_hash: Option<Digest>,
}

impl From<CertificateHeaderV0> for CertificateHeader {
    fn from(certificate: CertificateHeaderV0) -> Self {
        let CertificateHeaderV0 {
            network_id,
            height,
            epoch_number,
            certificate_index,
            certificate_id,
            prev_local_exit_root,
            new_local_exit_root,
            metadata,
            status,
            settlement_tx_hash,
        } = certificate;

        CertificateHeader {
            network_id,
            height,
            epoch_number,
            certificate_index,
            certificate_id,
            prev_local_exit_root,
            new_local_exit_root,
            metadata,
            status,
            settlement_tx_hash,
        }
    }
}

fn decode<T: for<'de> Deserialize<'de> + Into<CertificateHeader>>(
    bytes: &[u8],
) -> Result<CertificateHeader, CodecError> {
    Ok(default_bincode_options().deserialize::<T>(bytes)?.into())
}

impl crate::columns::Codec for CertificateHeader {
    fn encode(&self) -> Result<Vec<u8>, CodecError> {
        todo!()
    }

    fn decode(bytes: &[u8]) -> Result<Self, CodecError> {
        match bytes.first().copied() {
            None => Err(CodecError::CertificateEmpty),
            Some(0) => decode::<CertificateHeaderV0>(bytes),
            Some(version) => Err(CodecError::BadCertificateVersion { version }),
        }
    }
}
