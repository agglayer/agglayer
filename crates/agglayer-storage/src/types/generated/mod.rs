// @generated
pub mod agglayer {
    pub mod storage {
        // @@protoc_insertion_point(attribute:agglayer.storage.v0)
        pub mod v0 {
            include!("agglayer.storage.v0.rs");
            // @@protoc_insertion_point(agglayer.storage.v0)
        }
    }
}
