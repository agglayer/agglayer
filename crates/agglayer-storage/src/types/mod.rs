use agglayer_types::{
    primitives::Digest, CertificateHeader, CertificateId, CertificateIndex, EpochNumber, Height, NetworkId, Proof
};
use serde::{Deserialize, Serialize};

use crate::columns::Codec;

mod certificate;

macro_rules! default_codec_impl {
    ($($ident: ident),+) => {
        $(impl crate::columns::Codec for $ident {})+
    };
}

#[derive(Debug, Serialize, Deserialize)]
pub enum MetadataKey {
    LatestSettledEpoch,
    EpochSynchronization, // Actually unused, kept for storage backward compatibility
}

#[derive(Debug, Serialize, Deserialize)]
pub enum MetadataValue {
    LatestSettledEpoch(EpochNumber),
    EpochSynchronization(u64), // Actually unused, kept for storage backward compatibility
}

#[derive(Debug, Serialize, Deserialize)]
pub enum PerEpochMetadataKey {
    SettlementTxHash,
    Packed,
}

#[derive(Debug, Serialize, Deserialize)]
pub enum PerEpochMetadataValue {
    SettlementTxHash(Digest),
    Packed(bool),
}

#[derive(Debug, Serialize, Deserialize, PartialEq, Eq, PartialOrd, Ord)]
pub struct SmtKey {
    pub(crate) network_id: u32,
    pub(crate) key_type: SmtKeyType,
}

#[derive(Clone, Debug, Serialize, Deserialize, PartialEq, Eq, PartialOrd, Ord)]
pub enum SmtKeyType {
    Root,
    Node(Digest),
}

#[derive(Clone, PartialEq, Eq, Debug, Serialize, Deserialize)]
pub enum SmtValue {
    Node(Digest, Digest),
    Leaf(Digest),
}

impl Codec for SmtKey {}
impl Codec for SmtValue {}

default_codec_impl!(
    u64,
    u32,
    CertificateId,
    CertificateIndex,
    CertificateHeader,
    Digest,
    Height,
    MetadataKey,
    MetadataValue,
    NetworkId,
    PerEpochMetadataKey,
    PerEpochMetadataValue,
    Proof
);
