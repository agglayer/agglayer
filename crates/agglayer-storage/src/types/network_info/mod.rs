use std::io;

use prost::{bytes::BytesMut, Message};
use serde::{Deserialize, Serialize};
use strum::IntoEnumIterator;

pub use super::generated::agglayer::storage::v0;
use crate::{
    columns::Codec,
    types::network_info::v0::{
        network_info_value::{self, ValueDiscriminants},
        NetworkType,
    },
};

#[derive(Clone, Serialize, Deserialize)]
pub struct Key {
    pub(crate) network_id: u32,
    pub(crate) kind: ValueDiscriminants,
}

impl Key {
    pub(crate) fn all_keys_for_network(
        network_id: agglayer_types::NetworkId,
    ) -> impl ExactSizeIterator<Item = Self> + Clone {
        let network_id = network_id.to_u32();
        network_info_value::ValueDiscriminants::iter().map(move |kind| Key { network_id, kind })
    }
}

pub type Value = super::generated::agglayer::storage::v0::NetworkInfoValue;

impl Codec for Value {
    fn encode(&self) -> Result<Vec<u8>, crate::columns::CodecError> {
        let len = self.encoded_len();

        let mut buf = BytesMut::new();
        buf.reserve(len);
        <Value as prost::Message>::encode(self, &mut buf)?;

        Ok(buf.to_vec())
    }

    fn encode_into<W: io::Write>(&self, _writer: W) -> Result<(), crate::columns::CodecError> {
        unimplemented!()
    }

    fn decode(buf: &[u8]) -> Result<Self, crate::columns::CodecError> {
        <Value as prost::Message>::decode(buf).map_err(Into::into)
    }
}

impl TryFrom<v0::NetworkType> for agglayer_types::NetworkType {
    type Error = crate::columns::CodecError;

    fn try_from(value: v0::NetworkType) -> Result<Self, Self::Error> {
        match value {
            NetworkType::Generic => Ok(agglayer_types::NetworkType::Generic),
            NetworkType::Unspecified => Ok(agglayer_types::NetworkType::Unspecified),
            NetworkType::Ecdsa => Ok(agglayer_types::NetworkType::Ecdsa),
            NetworkType::MultisigOnly => Ok(agglayer_types::NetworkType::MultisigOnly),
            NetworkType::MultisigAndAggchainProof => {
                Ok(agglayer_types::NetworkType::MultisigAndAggchainProof)
            }
        }
    }
}

#[cfg(test)]
mod tests {
    use strum::EnumCount;

    use crate::types::network_info::v0::network_info_value::ValueDiscriminants;

    #[test]
    fn test_discriminant_from_u32() {
        assert_eq!(ValueDiscriminants::COUNT, 4, "Expected 4 discriminants");
        for i in 0..ValueDiscriminants::COUNT {
            let discriminant = ValueDiscriminants::from_repr(i);
            assert!(discriminant.is_some());

            match discriminant.unwrap() {
                ValueDiscriminants::NetworkType => {
                    assert_eq!(i, 0, "NetworkType should be at index 0")
                }
                ValueDiscriminants::SettledCertificate => {
                    assert_eq!(i, 1, "SettledCertificate should be at index 1")
                }
                ValueDiscriminants::SettledClaim => {
                    assert_eq!(i, 2, "SettledClaim should be at index 2")
                }
                ValueDiscriminants::LatestPendingCertificateInfo => {
                    assert_eq!(i, 3, "LatestPendingCertificateInfo should be at index 3")
                }
            }
        }
    }
}
