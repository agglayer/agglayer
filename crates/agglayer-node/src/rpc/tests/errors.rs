//! Tests for rendering RPC errors

use std::time::Duration;

use agglayer_contracts::polygon_rollup_manager::PolygonRollupManagerErrors;
use agglayer_types::Digest;
use alloy::{primitives::SignatureError as AlloySignatureError, signers::k256};
use ethers::{
    providers::ProviderError,
    types::{Bytes, SignatureError as EthSignatureError, H160, H256},
};
use jsonrpsee::types::ErrorObjectOwned;

use crate::{
    kernel::{self, ZkevmNodeVerificationError},
    rate_limiting::{self, component, Component},
    rpc::Error,
    service,
};

type RpcProvider = ethers::providers::Provider<ethers::providers::Http>;
type CheckTxStatusError = kernel::CheckTxStatusError<RpcProvider>;
type ContractError = ethers::contract::ContractError<RpcProvider>;
type SendTxError = service::SendTxError<RpcProvider>;
type SettlementError = kernel::SettlementError<RpcProvider>;
type SignatureError = kernel::SignatureVerificationError<RpcProvider>;
type TxStatusError = service::TxStatusError<RpcProvider>;
type WallClockLimitedInfo = <component::SendTx as Component>::LimitedInfo;

#[rstest::rstest]
#[case("rollup_not_reg", SendTxError::RollupNotRegistered { rollup_id: 1337 })]
#[case(
    "sig_invalid_len",
    SendTxError::SignatureError(SignatureError::CouldNotRecoverTxSigner(
        EthSignatureError::InvalidLength(42)
    ))
)]
#[case("sig_verif", SendTxError::SignatureError(SignatureError::CouldNotRecoverTxSigner(
    EthSignatureError::VerificationError(H160([0x11; 20]), H160([0x22; 20]))
)))]
#[case(
    "sig_recov",
    SendTxError::SignatureError(SignatureError::CouldNotRecoverTxSigner(
        EthSignatureError::RecoveryError
    ))
)]
#[case(
    "cert_sig",
    SendTxError::SignatureError(SignatureError::CouldNotRecoverCertSigner(
        AlloySignatureError::K256(k256::ecdsa::Error::new())
    ))
)]
#[case(
    "sig_signer",
    SendTxError::SignatureError(SignatureError::InvalidSigner{
        signer: H160([0x33; 20]),
        trusted_sequencer: H160([0x44; 20]),
    })
)]
#[case("sig_contract", SendTxError::SignatureError(ContractError::ContractNotDeployed.into()))]
#[case("dry_run_rollup_man", SendTxError::DryRunRollupManager(
        PolygonRollupManagerErrors::RevertString("Reverting".into())
))]
#[case(
    "root_bad_rollup",
    SendTxError::RootVerification(ZkevmNodeVerificationError::InvalidRollupId(13))
)]
#[case(
    "root_rpc",
    SendTxError::RootVerification(ZkevmNodeVerificationError::RpcError(
        jsonrpsee::core::client::Error::Custom("Node smells too much".into())
    ))
)]
#[case(
    "root_state",
    SendTxError::RootVerification(ZkevmNodeVerificationError::InvalidStateRoot {
        expected: H256([0x55; 32]),
        got: H256([0x66; 32]),
    })
)]
#[case(
    "root_exit",
    SendTxError::RootVerification(ZkevmNodeVerificationError::InvalidExitRoot {
        expected: H256([0x77; 32]),
        got: H256([0x88; 32]),
    })
)]
#[case("settle_receipt", SendTxError::Settlement(SettlementError::NoReceipt))]
#[case(
    "settle_io",
    SendTxError::Settlement(SettlementError::ProviderError(ProviderError::UnsupportedRPC))
)]
#[case(
    "settle_contract",
    SendTxError::Settlement(SettlementError::ContractError(ContractError::Revert(
        Bytes::from_static(b"foo")
    )))
)]
#[case(
    "settle_l1_timeout",
    SendTxError::Settlement(SettlementError::Timeout(Duration::from_secs(30 * 60)))
)]
#[case(
    "rate_disallowed",
    SendTxError::RateLimited(rate_limiting::RateLimited::SendTxDiabled {})
)]
#[case(
    "rate_sendtx",
    SendTxError::RateLimited(rate_limiting::RateLimited::SendTxRateLimited(WallClockLimitedInfo {
        max_per_interval: 3,
        time_interval: Duration::from_secs(30 * 60),
        until_next: Some(Duration::from_secs(123)),
    }))
)]
#[case(
    "rate_sendtx_nonext",
    SendTxError::RateLimited(rate_limiting::RateLimited::SendTxRateLimited(WallClockLimitedInfo {
        max_per_interval: 4,
        time_interval: Duration::from_secs(40 * 60),
        until_next: None,
    }))
)]
#[case(
    "txstatus_notfound",
    TxStatusError::TxNotFound { hash: H256([0x97; 32]) }
)]
#[case(
    "txstatus_check",
    TxStatusError::StatusCheck(CheckTxStatusError::ProviderError(
        ProviderError::SignerUnavailable
    ))
)]
#[case(
    "cert_notfound",
    service::CertificateRetrievalError::NotFound { certificate_id: Digest([0x51; 32]) }
)]
fn rpc_error_rendering(#[case] name: &str, #[case] err: impl Into<Error>) {
    let err: Error = err.into();
    let debug_str = format!("{err:?}");
    let err_obj = ErrorObjectOwned::from(err);
    let err_json_string = {
        // Going through an extra encode/decode here helps normalize the output.
        let json_string = serde_json::to_string(&err_obj).unwrap();
        let json = serde_json::from_str::<serde_json::Value>(&json_string).unwrap();
        serde_json::to_string_pretty(&json).unwrap()
    };

    insta::assert_snapshot!(name, err_json_string, &debug_str);
}
