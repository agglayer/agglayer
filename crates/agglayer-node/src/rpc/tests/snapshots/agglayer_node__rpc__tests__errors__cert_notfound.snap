---
source: crates/agglayer-node/src/rpc/tests/errors.rs
expression: "ResourceNotFound(\"Certificate(0x5151515151515151515151515151515151515151515151515151515151515151)\")"
snapshot_kind: text
---
{
  "code": -10008,
  "data": {
    "resource-not-found": "Certificate(0x5151515151515151515151515151515151515151515151515151515151515151)"
  },
  "message": "Resource not found: Certificate(0x5151515151515151515151515151515151515151515151515151515151515151)"
}
